Add a check for the leading and trailing spaces

amieiro opened the pull request:
As I described in https://github.com/WordPress/WordPress-Coding-Standards/issues/2501, at translate.w.org, sometimes we get translations that start or end with one or more empty spaces, and we want to avoid this.

This PR adds a new check to detect if a translatable string has leading or trailing spaces.

Fixes https://github.com/WordPress/WordPress-Coding-Standards/issues/2501.

amieiro committed:
Add the check for the leading and trailing spaces

amieiro committed:
Remove a debug code

amieiro committed:
Lint

jrfnl reviewed (commented):
Hi @amieiro, thanks for this PR.

I haven't tested it yet, but based on a code review alone, this is looking good.

I wonder if the tests can be made a little more varied ? I.e. not just use leading/trailing _spaces_, but also have some leading/trailing tabs, new lines and combinations of spaces, tabs, new lines.

What do you think ?

jrfnl commented:
@amieiro Just checking in as you never responded to the question I posed in my initia review.... Do you still intend to finish off this PR ?

amieiro committed:
Add tabs to the checks

amieiro committed:
Add new lines to the checks

amieiro committed:
Refactor the method to have different error messages for each check

amieiro committed:
Add vertical tabs to the checks

amieiro committed:
Change the regexs to use hex codes and reorganize some code

jrfnl commented:
@amieiro Seeing as you've been working on this again - could you please explain why you have set things up to have separate errors for spaces vs tabs vs vtabs vs newlines ?

My previous question was only about having tests in place with these and the sniff handling all types of whitespace, not about splitting that up into multiple errors.

I cannot think of any situation where one type of (leading/trailing) whitespace would be acceptable, but another would not be, so having different error codes doesn't make sense to me.

Is there a reason for this distinction based on your experiences in the Polyglots team ? I'd be interested to hear your reasoning for this.

jrfnl commented:
> @jrfnl Sorry, this PR is not yet ready for review and I forgot to convert it to draft. As we [return to contribute](https://automattic.com/2025/05/29/returning-to-core/) with WordPress.org, I resumed work on this PR.

Okay, thanks for letting me know. I've converted it to draft now and won't do an in-detail review until it has been marked as "Ready to review".


> The horizontal tabulations are not often seen at the beginning or end of sentences, but they are seen in the middle, so I wanted to prevent its use, since I don't think it is correct to use it. You can see a use case inside a string in the core [here](https://translate.wordpress.org/projects/wp/dev/gl/default/?filters%5Bstatus%5D=either&filters%5Boriginal_id%5D=9144351&filters%5Btranslation_id%5D=70644590). The same applies to the new lines. You can see an example in the core [here](https://translate.wordpress.org/projects/wp/dev/gl/default/?filters%5Bstatus%5D=either&filters%5Boriginal_id%5D=5143989&filters%5Btranslation_id%5D=52619651).

I think that this PR should probably focus on the original problem it was trying to solve - leading and trailing whitespace.

Expanding the scope to mid-string whitespace characters feels like scope-creep and may have different implications (i.e., would involve a different decision point, potentially delaying the merging of the code related to the original problem).

I suggest opening a follow-up issue for that to discuss if this needs flagging and if so, what would need flagging. (I mean, looking at the examples, these seem like perfectly valid use-cases for mid string vertical whitespace, especially if you also consider RTL and BTT languages).

> And we have another thing to check: the different whitespace characters. The regular space characters is U+0020, but there are a lot of different [whitespace characters](https://en.wikipedia.org/wiki/Whitespace_character#Unicode): U+00A0, U+2000, U+2001, U+2002,...
> 
> The most common problem is the U+0020 space problem (in fact it is the only one I usually see), so the others are really edge situations. I can continue this PR in two different approaches:
> 
> * Contemplate only the U+0020 case and add the other cases to the tests.
> * Add all the commented situations.
> 
> Which approach do you prefer?

For leading and trailing whitespace, IMO the new errors should flag all whitespace characters, but without differentiating in the error message or code.
What I mean by that is:
* Change the error message(s) to use the term "whitespace" instead of explicitly saying what type of whitespace you're flagging.
* Change the error code(s) to also use the term "whitespace" - this would mean you'd end up with two error codes: `LeadingWhiteSpace` and `TrailingWhiteSpace`.

Does that help ?

jrfnl commented:
@amieiro Just checking in - will you have time to finish this PR in the near future ? It's been over a year since you opened it and it's still not marked as ready for review, while the base premise of the PR seems like a valid and valuable addition, which would help WPCS users.

amieiro committed:
♻️ refactor(I18n): consolidate leading/trailing whitespace check into two error codes

- replace eight per-character error codes (Leading/Trailing x Spaces/Tabs/VTabs/NewLines)
  with two generic codes: LeadingWhiteSpace and TrailingWhiteSpace
- reduce the eight hex-code regexes to a single `\s` match on each side (covers space,
  tab, vertical tab, newline, CR, form feed in one pass)
- reword error messages to use the generic term "whitespace"
- bump @since annotation from 3.2.0 to 3.4.0
- add four combination test cases (leading tab+space, trailing space+tab, mixed
  leading/trailing, multi-line with newline+tab on each side) to verify strings with
  multiple whitespace types produce at most one leading and one trailing error

rodrigoprimo commented on WordPress/Sniffs/WP/I18nSniff.php:
```diff
@@ -804,6 +805,42 @@ private function check_string_has_no_html_wrapper( $matched_content, $param_name
+		$content_without_quotes = TextStrings::stripQuotes( $param_info['clean'] );
+		$first_non_empty        = $this->phpcsFile->findNext( Tokens::$emptyTokens, $param_info['start'], ( $param_info['end'] + 1 ), true );
+
+		if ( preg_match( '/^\s+/u', $content_without_quotes ) ) {
```
Nitpick: I believe the `+` quantifier is redundant here and in the regex below. A single surrounding whitespace character is enough to trigger these errors.
```suggestion
		if ( preg_match( '/^\s/u', $content_without_quotes ) ) {
```

rodrigoprimo replied:
I wonder about the use of the `u` modifier. I assume it was added so that `\s` also matches non-ASCII whitespace characters? I worry a bit about unintended consequences of this modifier, as PHP's UTF-8 handling can be subtle, but I'm not sure whether any such edge cases apply here.

If it is important to match non-ASCII whitespace, I'd suggest adding tests to cover the cases the sniff should catch (presumably the ones you mentioned [here](https://github.com/WordPress/WordPress-Coding-Standards/pull/2504#issuecomment-4297916642)). Currently, the tests pass if the `u` modifier is removed.

rodrigoprimo commented on WordPress/Sniffs/WP/I18nSniff.php:
```diff
@@ -804,6 +805,42 @@ private function check_string_has_no_html_wrapper( $matched_content, $param_name
+	 *
+	 * @return void
+	 */
+	private function check_string_has_no_leading_trailing_whitespace( $matched_content, $param_name, $param_info ) {
```
This method accepts `$matched_content` and `$param_name` but doesn't use either. The sibling `check_string_has_no_html_wrapper()` and other related methods (introduced in c9e8dbd0) have the same pattern.

@jrfnl, is there a reason for the uniform signature with unused parameters that I'm missing?

If not, I lean toward dropping the unused parameters for `check_string_has_no_leading_trailing_whitespace()`.

jrfnl replied:
> @jrfnl, is there a reason for the uniform signature with unused parameters that I'm missing?

IIRC, in the original sniff (before extensive refactoring), most, if not all, the "sub-functions" were called via `call_user_func()`, which meant the expected parameters of these "sub-functions" needed to be the same.

You may very well be correct in that the various refactoring done since that time have made that requirement redundant. This would need to be verified and confirmed and if so, any method signatures for `private` methods can be cleaned up.

rodrigoprimo replied:
Thanks for the additional details, @jrfnl.

@amieiro, considering the above, I think it is better to remove the unused parameters from `check_string_has_no_leading_trailing_whitespace()`. No need to change the other methods.

rodrigoprimo commented on WordPress/Sniffs/WP/I18nSniff.php:
```diff
@@ -804,6 +805,42 @@ private function check_string_has_no_html_wrapper( $matched_content, $param_name
+	 *
+	 * @return void
+	 */
+	private function check_string_has_no_leading_trailing_whitespace( $matched_content, $param_name, $param_info ) {
```
Nitpick: maybe `check_string_has_no_surrounding_whitespace()` is a clearer name?

rodrigoprimo commented on WordPress/Tests/WP/I18nUnitTest.1.inc:
```diff
@@ -328,4 +328,56 @@ MyNamespace\__( 'foo', 'my-slug' ); // Ok.
 namespace\esc_html_e( 'foo', '' ); // The sniff should start flagging this once it can resolve relative namespaces.
 namespace\Sub\translate( 'foo', 'my-slug' ); // Ok.
 
+__( ' string', 'my-slug' ); // Bad: leading space.
```
Looking at the new test cases for this check, I wonder if there is a reason to have duplicate tests that differ only in function name? They seem redundant to me because the method's logic doesn't vary based on the function name.

To me, the tests should cover different whitespace positions, whitespace types, and potentially parameter names, plus a few variations in whitespace quantity to guard against future regressions. What do you think?

rodrigoprimo reviewed (commented):
Thanks for updating this PR, @amieiro. I confirmed that you addressed the points raised in previous reviews. Besides that, I also left some inline comments for us to discuss.
